### ISSUE TITLE
External persistence

### ISSUE THREAD

**Issue by NMinhNguyen** [bvaughn/react-resizable-panels#41]
Would it be possible to expose a way to provide custom persistence (for example, if you wanna persist the sizes/layouts to a database rather than local storage)?

**Comment by bvaughn** [comment]
@NMinhNguyen Yes, it would be possible to add support for that, but it would add a little complexity (to the implementation and more importantly to the API) so I'd like to better understand the use case.

**Comment by NMinhNguyen** [comment]
The use case is being able to log in from another device and having the layouts restored from external persistence. I wonder if it could instead be achieved via supplying `defaultSize`, except I don't see any `onChange`-style APIs that would let me keep track of the sizes.

**Comment by bvaughn** [comment]
> I wonder if it could instead be achieved via supplying `defaultSize`

I think this seems like a good compromise, if there was a "layout changed" event.

**Comment by NMinhNguyen** [comment]
> I think this seems like a good compromise, if there was a "layout changed" event.

Agreed! But don't feel like you have to implement this straight away. I'd rather you didn't come up with an API that can cause maintenance issues just to satisfy this use case.

**Comment by bvaughn** [comment]
I appreciate you saying that 😌

The thing that's giving me pause is that I don't want to expose anything about _how_ the size information is persisted. (In order to support conditional panels, `PanelGroup` has to potentially serialize multiple size configurations.)

Since `Panel` is the thing that would accept a `defaultSize` param, I think it would probably make the most sense to add an `onChange` or `onResize` to that component. That might be a little awkward _for you_, since you'd have to listen to multiple `Panel` components individually, but I think that would be the most future-proof for the API itself in terms of not exposing internal data structures.

**Comment by bvaughn** [comment]
If I went that route, you'd use something like this for local out-of-the-box persistence:
```jsx
<PanelGroup autoSaveId="persistence" direction="horizontal">
  <Panel>
    {/* ... */}
  </Panel>
  <ResizeHandle />
  <Panel>
        {/* ... */}
  </Panel>
```

**Comment by bvaughn** [comment]
If I went that route, you'd use something like this for local out-of-the-box persistence:
```jsx
<PanelGroup autoSaveId="persistence" direction="horizontal">
  <Panel>
    {/* ... */}
  </Panel>
  <ResizeHandle />
  <Panel>
        {/* ... */}
  </Panel>
</PanelGroup>
```

And you'd use something like this for remote persistence:
```jsx
<PanelGroup direction="horizontal">
  <Panel defaultSize={sizeOne} onResize={saveSizeOne}>
    {/* ... */}
  </Panel>
  <ResizeHandle />
  <Panel defaultSize={sizeTwo} onResize={saveSizeTwo}>
        {/* ... */}
  </Panel>
</PanelGroup>
```

**Comment by NMinhNguyen** [comment]
Sounds good @bvaughn! :) I suppose another thing (although super hacky) I could do right now is seed `localStorage` with the right data before mounting `PanelGroup` but that's super brittle lol

**Comment by bvaughn** [comment]
Yeah, that would be pretty brittle.

**Comment by bvaughn** [comment]
Released in v0.0.23

**Comment by ozgurrgul** [comment]
I feel like instead of providing autoSaveId to each `PanelGroup`, having an api like this is better:

``` tsx
const [tmpDimensions, setTmpDimensions]= useState(...);

<PanelOrchestration // <-- new wrapper
  onChange={(dimensions) => setTmpDimensions(dimensions)} 
  dimensions={tmpDimensions}
>

  <PanelGroup direction="horizontal">
    <Panel>{/* ... */}</Panel>
    <ResizeHandle />
    <Panel>{/* ... */}</Panel>
  </PanelGroup>

</PanelOrchestration>
```

If you can't imperatively provide initial dimensions to the panels, then it doesn't make much sense to have an `onResize` callback for each panel group (as it's now). Simple listen to localStorage events in window since it's hard dependency.

Another way to make it more extendable by end users is to provide a `storageApi`. It can be `window.localStorage` or any other custom implementation I assume. (Just brain storming since I am working on similar component)
